src/librustc/hir/mod.rs

```diff
@@ -444,71 +446,54 @@ pub enum TraitBoundModifier {
 pub type TyParamBounds = HirVec<TyParamBound>;
 
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct TyParam {
-    pub name: Name,
+pub enum GenericParamKind {
+    /// A lifetime definition, eg `'a: 'b + 'c + 'd`.
+    Lifetime {
+        /// Either "'a", referring to a named lifetime definition,
+        /// or "" (aka keywords::Invalid), for elision placeholders.
+        ///
+        /// HIR lowering inserts these placeholders in type paths that
+        /// refer to type definitions needing lifetime parameters,
+        /// `&T` and `&mut T`, and trait objects without `... + 'a`.
+        name: LifetimeName,
+        bounds: HirVec<Lifetime>,
+        // Indicates that the lifetime definition was synthetically added
+        // as a result of an in-band lifetime usage like:
+        // `fn foo(x: &'a u8) -> &'a u8 { x }`
+        in_band: bool,
+        // We keep a `Lifetime` around for now just so we can `visit_lifetime`.
+        lifetime: Lifetime,
+    },
+    Type {
+        name: Name,
+        bounds: TyParamBounds,
+        default: Option<P<Ty>>,
+        synthetic: Option<SyntheticTyParamKind>,
+        attrs: HirVec<Attribute>,
```

Comment:
Lifetime parameters have attributes too, where do those go? In `Lifetime`? That doesn't seem right.

Reply:
`ast::GenericParam` has attributes for both lifetime and type parameters, but they don't seem to persist to HIR (the previous definition [didn't have them either](https://github.com/rust-lang/rust/pull/48149/files/4c62185044b4d714f942a94e05da17baacce422d#diff-7fceb7ede15b205bf5ad812c31d75384L278)).